### PR TITLE
refactored to TypeScript & bug fixes

### PR DESCRIPTION
**Pull request by ilgooz**
closes https://github.com/mesg-foundation/mesg-js/issues/7, https://github.com/mesg-foundation/mesg-js/issues/9, https://github.com/mesg-foundation/mesg-js/issues/12, https://github.com/mesg-foundation/mesg-js/issues/21

## Log
- [x] add unit tests before a potential merge.
- [x] works as expected when tested with [Quick Start Guide](https://docs.mesg.com/start-here/quick-start-guide#3-deploy-the-services) example, no breaking API changes.
- [x] refactored to TypeScript with OO programming style.
- [x] application spesific types added.
- [x] added types for proto files.
- [x] type safety provided for both internal development of `mesg-js` and for developers who install `mesg-js` via npm.
- [x] minor bug fixes.
- [x] switched from `grpc.load` to `grpc.loadPackageDefinition` due to deprecation warning.
- [x] `npm run deploy` command added to publish npm.

## Sample usage with TS syntax:
image

Here is a Gopher, totally irrelevant

image

### CODE REVIEW

**Comment by ilgooz** [comment]
@NicolasMahe yes, I crafted proto typings by hand. I was going to explain some details about why chose it but I thought you made the decision before by using dynamically generated protos on runtime.

I share the same opinion as you are but I didn't like how [ts-protoc-gen](https://github.com/improbable-eng/ts-protoc-gen) and [grpc_tools_node_protoc_ts ](https://github.com/agreatfool/grpc_tools_node_protoc_ts)generates proto files for javascript. Because they require using setters for our data types as you can see it here:
* https://github.com/agreatfool/grpc_tools_node_protoc_ts/blob/master/examples/src/client.ts#L15-L16
* https://github.com/improbable-eng/ts-protoc-gen#generating-typescript-definitions-for-commonjs-modules-generated-by-protoc

This behavior also causes some breaking changes in our API since we can't directly pass data objects like this:
* https://github.com/mesg-foundation/mesg-js#execute-a-task

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    mesgConfig: mesgConfig,
+    client: new ClientBuilder({
+      endpoint: endpointTCP,
+    }).service(),
+  });
+}
+
+const application = () => {
+  return new Application({
+    token: token,
```
Comment (antho1404) [inline src/index.ts]: `token` is not useful in the application, this is just something used to secures the connection between the core and the service

File: src/application/application.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        await this.startService(result.serviceID);
+        await this.startService(task.serviceID);
+
+        const stream = this.client.listenResult({
+            serviceID: result.serviceID,
+            taskFilter: result.task || '*',
+            outputFilter: result.output || '*'
+        });
+        stream.on('data', async ({ outputKey, outputData }) => {
+            await this.executeTask(task, outputKey, outputData);
```
Comment (NicolasMahe) [inline src/application/application.ts]: Why waiting for the Promise here? And not there: https://github.com/mesg-foundation/mesg-js/pull/20/files#diff-bc3351f31e4b582dafa2dc8531df0fa0R35

[the remark this answers is not in the document]
Comment (ilgooz) [reply src/application/application.ts]: Actually that's a forgotten unhandled promise here https://github.com/ilgooz/mesg-js/blob/0e3383428d1f9201d78885084686b8edd524060d/src/application/application.ts#L35. I think unhandled promises will not throw an error in future versions of v8 so it's a good idea to but an await.

Either way gRPC 'data' event's callback does not react on any thrown errors. So we maybe want to fire an 'error' event on stream for user to handle this.
https://grpc.io/docs/tutorials/basic/node.html#streaming-rpcs

File: src/client/client.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+class ClientBuilder {
+    private options: Options;
+
+    constructor(options: Options){
+        this.options = options;
+    }
+
+    service(): ServiceClient {
+        const packageDefinition = protoLoader.loadSync(path.join(__dirname, './proto/api-service.proto'));
```
[the remark this answers is not in the document]
Comment (ilgooz) [reply src/client/client.ts]: 👍

File: src/client/client.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      
+        const clientConstructor = packageObject.api.Service;
+        return <ServiceClient>new clientConstructor(
+            this.options.endpoint,
+            grpc.credentials.createInsecure()
+        )
+    }
+
+    core(): CoreClient {
+        const packageDefinition = protoLoader.loadSync(path.join(__dirname, './proto/api-core.proto'));
```
[the remark this answers is not in the document]
Comment (ilgooz) [reply src/client/client.ts]: 👍

**Pull request merged**